narwhal/types/src/pre_subscribed_broadcast.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use tokio::sync::broadcast;
+use tokio::sync::broadcast::error::SendError;
+
+/// PreSubscribedBroadcastSender is a wrapped Broadcast channel that limits
+/// subscription to initialization time. This is designed to be used for cancellation
+/// signal to all the components, and the limitation is intended to prevent a component missing
+/// the shutdown signal due to a subscription that happens after the shutdown signal was sent.
+/// The receivers have a special peek method which can be used to conditionally check for
+/// shutdown signal on the channel.
+pub struct PreSubscribedBroadcastSender {
+    sender: broadcast::Sender<()>,
+    receivers: Vec<ConditionalBroadcastReceiver>,
+}
+
+pub struct ConditionalBroadcastReceiver {
+    pub receiver: broadcast::Receiver<()>,
+}
+
+/// ConditionalBroadcastReceiver has an additional method for convenience to be able to use
+/// to conditionally check for shutdown in all branches of a select statement. Using this method
+/// will allow for the shutdown signal to propagate faster, sice we will no longer be waiting
+/// until the branch that checks the receiver is randomly selected by the select macro.
+impl ConditionalBroadcastReceiver {
+    pub async fn received_signal(&mut self) -> bool {
```

Comment:
I see that this method is only used in tests, can you give some example how it can be used?

Reply:
This will be used to optimize the time it takes for shutdown, and is a syntactic sugar for being able to conditionally check for shutdown anywhere in our code. It can be used inside the first block of code of other branches of select statements, so that it has the effect of overriding the either completely random or completely biased selection of select branches. By using that conditional check, even if we randomly selected another future in the select, we can check if the shutdown future is ready, and if so, return instead of doing more work before shutting down.